packages/react/src/jsx.tsx

```diff
@@ -49,24 +48,31 @@ export function jsx(
   }
 }
 
-export type Extend<T> = { [P in keyof T]?: _Extend<T[P]> }
+export type Only<T> = {
+  __only: T
+}
+export type WithServerProps<T> = { [P in keyof T]: _Extend<T[P]> }
 
-// eslint-disable-next-line @typescript-eslint/ban-types
-export type _Extend<T> = T extends Function
-  ? T | ServerEventDataState | ActionDataState
-  : T extends Array<infer U>
-    ? _DeepPartialArray<U>
-    : T extends object
-      ? Extend<T>
-      : T | ReferencedDataState | StateDataState
+export type _Extend<T> =
+  T extends Only<infer U>
+    ? U
+    : T extends (...args: any) => any
+      ? T | ServerHandlerDataState
+      : T extends Array<infer U>
+        ? _DeepPartialArray<U>
+        : T extends object
+          ? WithServerProps<T>
+          : T extends null | undefined
```

Comment:
improvement: if prop is null or undefined, leave it as is. Previously, it would turn it into "StateDataState | null", so if there was the following type:
```
type Props = {
  onPress?: Function
}
```
it would turn it into
```
{
   onPress: Function | ServerHandlerDataState | ReferencedDataState | StateDataState | undefined
}
```
where in fact, we want
```
{
   onPress: Function | ServerHandlerDataState | undefined
}
```